### PR TITLE
Log domain/workflow/run ids for non-deterministic error from RespondD…

### PR DESCRIPTION
**Pull request by luowenking1**
…ecisionTaskFail (#2929)

### CODE REVIEW

File: service/history/handler.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ import (
 	"github.com/uber/cadence/.gen/go/history/historyserviceserver"
 	r "github.com/uber/cadence/.gen/go/replicator"
 	gen "github.com/uber/cadence/.gen/go/shared"
+	workflow "github.com/uber/cadence/.gen/go/shared"
```
Comment (yycptt) [inline service/history/handler.go]: L37 already imported the package.

File: common/metrics/defs.go [hunk 2]
```diff
@@ -1516,6 +1516,8 @@ const (
 	MatchingClientForwardedCounter
 	MatchingClientInvalidTaskListName
 
+	NonDeterministicErrorCount
```
Comment (yycptt) [inline common/metrics/defs.go]: Please add the metric to the `MetricDefs` map.

Comment (yycptt) [reply common/metrics/defs.go]: Please also group this metric with other error metrics starting from L1431.

File: service/history/handler.go [hunk 3]
```diff
@@ -550,6 +551,14 @@ func (h *Handler) RespondDecisionTaskFailed(
 		token.RunID,
 		token.ScheduleID))
 
+	if failedRequest != nil && failedRequest.GetCause() == workflow.DecisionTaskFailedCauseUnhandledDecision {
+		h.GetLogger().Info("Non-Deterministic Error", tag.WorkflowRunID(token.RunID))
```
Comment (yycptt) [inline service/history/handler.go]: Please also tag the domainID and workflowID

File: service/history/handler.go [hunk 4]
```diff
@@ -550,6 +551,14 @@ func (h *Handler) RespondDecisionTaskFailed(
...
+		domainName, err := h.GetDomainCache().GetDomainName(token.DomainID)
+		if err == nil {
+			nonDeterministic := h.GetMetricsClient().Scope(metrics.HistoryRespondDecisionTaskFailedScope, metrics.DomainTag(domainName))
+			nonDeterministic.IncCounter(metrics.NonDeterministicErrorCount)
```
Comment (yycptt) [inline service/history/handler.go]: `scope` is already defined on L527, you can reuse its value here. Also it's not necessary to create a temp variable `nonDeterministic`.

File: service/history/handler.go [hunk 5]
```diff
@@ -550,6 +551,14 @@ func (h *Handler) RespondDecisionTaskFailed(
 		token.RunID,
 		token.ScheduleID))
 
+	if failedRequest != nil && failedRequest.GetCause() == workflow.DecisionTaskFailedCauseUnhandledDecision {
+		h.GetLogger().Info("Non-Deterministic Error", tag.WorkflowRunID(token.RunID))
+		domainName, err := h.GetDomainCache().GetDomainName(token.DomainID)
+		if err == nil {
```
Comment (yycptt) [inline service/history/handler.go]: If err is not nil, you can emit the metric with `DomainUnknownTag`

File: common/metrics/defs.go [hunk 6]
```diff
@@ -1799,6 +1798,7 @@ var MetricDefs = map[ServiceIdx]map[int]metricDefinition{
 		CadenceErrBadBinaryCounter:                          {metricName: "cadence_errors_bad_binary", metricType: Counter},
 		CadenceErrClientVersionNotSupportedCounter:          {metricName: "cadence_errors_client_version_not_supported", metricType: Counter},
 		CadenceErrIncompleteHistoryCounter:                  {metricName: "cadence_errors_incomplete_history", metricType: Counter},
+		CadenceErrNonDeterministicCounter:                   {metricName: "cadence_NonDeterministic_Error", metricType: Counter},
```
Comment (vancexu) [inline common/metrics/defs.go]: Let's follow the pattern for metricName: cadence_errors_nondeterministic

File: service/history/handler.go [hunk 7]
```diff
@@ -551,12 +550,14 @@ func (h *Handler) RespondDecisionTaskFailed(
 		token.RunID,
 		token.ScheduleID))
 
-	if failedRequest != nil && failedRequest.GetCause() == workflow.DecisionTaskFailedCauseUnhandledDecision {
-		h.GetLogger().Info("Non-Deterministic Error", tag.WorkflowRunID(token.RunID))
+	if failedRequest != nil && failedRequest.GetCause() == gen.DecisionTaskFailedCauseUnhandledDecision {
+		h.GetLogger().Info("Non-Deterministic Error", tag.WorkflowDomainID(token.DomainID), tag.WorkflowID(token.WorkflowID), tag.WorkflowRunID(token.RunID))
 		domainName, err := h.GetDomainCache().GetDomainName(token.DomainID)
+		counter := metrics.CadenceErrNonDeterministicCounter
```
Comment (vancexu) [inline service/history/handler.go]: Maybe metricName is more proper than counter here? since `counter` make people think it's a number.

Comment (yycptt) [reply service/history/handler.go]: I would suggest create a variable for domainTag and set different for it based on the error returned. Then you can write  `h.GetMetricsClient().Scope(scope, tag).IncCounter(metrics.CadenceErrNonDeterministicCounter)` only once. Just my personal opinion. Your call.

File: service/history/handler.go [hunk 8]
```diff
@@ -550,6 +550,19 @@ func (h *Handler) RespondDecisionTaskFailed(
 		token.RunID,
 		token.ScheduleID))
 
+	if failedRequest != nil && failedRequest.GetCause() == gen.DecisionTaskFailedCauseUnhandledDecision {
```
Comment (yycptt) [inline service/history/handler.go]: Can you add a comment here which basically says we need a better way to determine if decision task failed due to non-deterministic error?

File: service/history/handler.go [hunk 9]
```diff
@@ -550,6 +550,19 @@ func (h *Handler) RespondDecisionTaskFailed(
...
+		h.GetLogger().Info("Non-Deterministic Error", tag.WorkflowDomainID(token.DomainID), tag.WorkflowID(token.WorkflowID), tag.WorkflowRunID(token.RunID))
```
Comment (yycptt) [inline service/history/handler.go]: Sorry, I forgot this in my previous comments. It's possible to also log the error details here?

**Pull request merged**